net/network.go

```diff
@@ -1,22 +1,72 @@
 package net
 
 import (
+	"context"
+	"fmt"
+	"sort"
+
 	ma "gx/ipfs/QmNTCey11oxhb1AxDnQBRHtdhap6Ctud872NjAYPYYXPuc/go-multiaddr"
 	"gx/ipfs/QmTu65MVbemtUxJEWgsTtzv9Zv9P8rvmqNA4eG9TrTRGYc/go-libp2p-peer"
+	"gx/ipfs/QmU7iTrsNaJfu1Rf5DrvaJLH9wJtQwmP4Dj8oPduprAU68/go-libp2p-swarm"
+	"gx/ipfs/QmVmDhyTTUcQXFD1rRQ64fGLMSAoaQvNH3hwuaCFAPq2hy/errors"
+	routing "gx/ipfs/QmWaDSNoSdSXU9b6udyaq9T8y6LkzMwqWxECznFqvtcTsk/go-libp2p-routing"
 	"gx/ipfs/QmZZseAa9xcK6tT3YpaShNUAEpyRAoWmUL5ojH3uGNepAc/go-libp2p-metrics"
 	"gx/ipfs/QmcNGX5RaxPPCYwa6yGXM1EcUbrreTTinixLcYGmMwf1sx/go-libp2p/p2p/protocol/ping"
 	"gx/ipfs/Qmd52WKRSwrBK5gUaJKawryZQ5by6UbNB8KVW2Zy6JtbyW/go-libp2p-host"
 
 	"github.com/filecoin-project/go-filecoin/net/pubsub"
 )
 
+// SwarmConnInfo represents details about a single swarm connection.
+type SwarmConnInfo struct {
+	Addr    string
+	Peer    string
+	Latency string
+	Muxer   string
+	Streams []SwarmStreamInfo
+}
+
+// SwarmStreamInfo represents details about a single swarm stream.
+type SwarmStreamInfo struct {
+	Protocol string
+}
+
+func (ci *SwarmConnInfo) Less(i, j int) bool {
+	return ci.Streams[i].Protocol < ci.Streams[j].Protocol
+}
+
+func (ci *SwarmConnInfo) Len() int {
+	return len(ci.Streams)
+}
+
+func (ci *SwarmConnInfo) Swap(i, j int) {
+	ci.Streams[i], ci.Streams[j] = ci.Streams[j], ci.Streams[i]
+}
+
+// SwarmConnInfos represent details about a list of swarm connections.
+type SwarmConnInfos struct {
+	Peers []SwarmConnInfo
+}
+
+func (ci SwarmConnInfos) Less(i, j int) bool {
+	return ci.Peers[i].Addr < ci.Peers[j].Addr
+}
+
+func (ci SwarmConnInfos) Len() int {
+	return len(ci.Peers)
+}
+
+func (ci SwarmConnInfos) Swap(i, j int) {
+	ci.Peers[i], ci.Peers[j] = ci.Peers[j], ci.Peers[i]
+}
+
 // Network is a unified interface for dealing with libp2p
 type Network struct {
 	host host.Host
 	*pubsub.Subscriber
 	*pubsub.Publisher
 	metrics.Reporter
-	*Router
+	Router routing.IpfsRouting
```

Comment:
Blocking: hey @anorth, this works for me but you blocked this when I first did it so wanted to run it by you to make sure you are on board.

Reply:
Yes, please indirect through the net.Router. The reason for this type is to abstract out libp2p/ipfs from most of the code so that we can more easy fake it for testing and/or upgrade the dependee code with minimal disruption elsewhere. I realise we have yet to clearly demonstrate how and the benefits of doing so in the rest of the codebase (but you can refer to pubsub for an example).

Thanks Zen.